pkg/kubelet/kubelet.go

```diff
@@ -3076,3 +3061,79 @@ func (kl *Kubelet) warnCgroupV1Usage() {
 	}
 	metrics.CgroupVersion.Set(float64(cgroupVersion))
 }
+
+// staticPodRegistration ensures that all static pods are registered to the apiserver
+// before node became ready.
+func (kl *Kubelet) initialStaticPodsRegistration() error {
+	kl.initialStaticPodsRegisteredLock.Lock()
+	defer kl.initialStaticPodsRegisteredLock.Unlock()
+	// kubelet running in standalone mode does not register static pods to the apiserver.
+	if kl.kubeClient == nil {
+		return nil
+	}
+	// Check if initial static pod registration has already been completed.
+	if kl.initialStaticPodsRegistered {
+		return nil
+	}
+	staticPodToMirrorPodMap := kl.podManager.GetStaticPodToMirrorPodMap()
+	for _, mirrorPod := range staticPodToMirrorPodMap {
+		if mirrorPod == nil {
+			return fmt.Errorf("not all static pods are registered")
+		}
+	}
+
+	kl.initialStaticPodsRegistered = true
+	return nil
+}
+
+// Ensure Mirror Pod for Static Pod exists.
+func (kl *Kubelet) ensureMirrorPodExists(staticPod, mirrorPod *v1.Pod) {
+	if kubetypes.IsStaticPod(staticPod) {
+		deleted := false
+		if mirrorPod != nil {
+			if mirrorPod.DeletionTimestamp != nil || !kubepod.IsMirrorPodOf(mirrorPod, staticPod) {
+				// The mirror pod is semantically different from the static pod. Remove
+				// it. The mirror pod will get recreated later.
+				klog.InfoS("Trying to delete pod", "pod", klog.KObj(staticPod), "podUID", mirrorPod.ObjectMeta.UID)
+				podFullName := kubecontainer.GetPodFullName(staticPod)
+				var err error
+				deleted, err = kl.mirrorPodClient.DeleteMirrorPod(podFullName, &mirrorPod.ObjectMeta.UID)
+				if deleted {
+					klog.InfoS("Deleted mirror pod because it is outdated", "pod", klog.KObj(mirrorPod))
+				} else if err != nil {
+					klog.ErrorS(err, "Failed deleting mirror pod", "pod", klog.KObj(mirrorPod))
+				}
+			}
+		}
+		if mirrorPod == nil || deleted {
+			node, err := kl.GetNode()
+			if err != nil {
+				klog.V(4).ErrorS(err, "No need to create a mirror pod, since failed to get node info from the cluster", "node", klog.KRef("", string(kl.nodeName)))
+			} else if node.DeletionTimestamp != nil {
+				klog.V(4).InfoS("No need to create a mirror pod, since node has been removed from the cluster", "node", klog.KRef("", string(kl.nodeName)))
+			} else {
+				klog.V(4).InfoS("Creating a mirror pod for static pod", "pod", klog.KObj(staticPod))
+				if err := kl.mirrorPodClient.CreateMirrorPod(staticPod); err != nil {
+					klog.ErrorS(err, "Failed creating a mirror pod for", "pod", klog.KObj(staticPod))
+				}
+			}
+		}
+	}
+}
+
+// Ensure Mirror Pod for Static Pod exists as soon as node is registered.
+func (kl *Kubelet) fastStaticPodsRegistration() {
```

Comment:
can you please add a context here ?
```suggestion
func (kl *Kubelet) fastStaticPodsRegistration(ctx context.Context) {
```